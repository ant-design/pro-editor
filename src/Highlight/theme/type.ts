export const THEME_DARK = 'dark';
export const THEME_LIGHT = 'light';

// 主题类型
const tuple = <T extends string[]>(...args: T) => args;
const ThemeTypes = tuple(THEME_DARK, THEME_LIGHT);
export type ThemeType = (typeof ThemeTypes)[number];
