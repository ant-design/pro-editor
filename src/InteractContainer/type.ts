import type { ReactNode } from 'react';

export type InteractionType = 'click' | 'hover';

// export enum InteractStatusEnum {
//   Hover = 'hover',
//   Unhover = 'unhover',
//   Selected = 'selected',
//   UnSelected = 'unSelected',
// }

export type InteractStatusType =
  | 'hover'
  | 'unhover'
  | 'unSelected'
  | 'selected';

// export type InteractStatusType = typeof InteractStatusEnum[keyof typeof InteractStatusEnum];
/**
 * 激活的节点
 */
export interface InteractStatusNode {
  index: number;
  rule: CanvasInteractRule;
}

/**
 * 交互规则
 */
export interface CanvasInteractRule {
  /**
   * 规则id
   */
  id: string;
  /**
   * 用于选择对象
   * 支持使用 类、id 选择器
   */
  selectors: string[];
  /**
   * 支持的交互行为
   */
  actions: InteractionType[];
}

export type ContainerNode = Document | Element;
/**
 * 交互模型
 */
export interface CanvasInteractModel extends CanvasInteractRule {
  elements: Element[];
  /**
   * 容器是否匹配
   */
  containerMatched: boolean;
}

export interface ValidElement {
  /**
   * 元素
   */
  element: HTMLElement;
  /**
   * 选择器
   */
  selector: string;
  /**
   * 序号
   */
  index: number;
}

export interface InteractStatus {
  /**
   * 交互状态的动作
   */
  action?: InteractionType;
  /**
   * 交互状态的动作
   */
  status?: InteractStatusType;
  /**
   * 当前的交互节点
   */
  currentNode?: InteractStatusNode;
  /**
   * 实际激活的节点
   */
  activeNode?: InteractStatusNode;
}

export type GetContainer = (() => HTMLElement) | HTMLElement | string;

export type OnInteractionStatusChange = (status: InteractStatus) => void;

export interface ContextCanvasProps {
  /**
   * 交互规则
   */
  rules: CanvasInteractRule[];
  /**
   * 受控模式的状态
   */
  status?: InteractStatus;
  /**
   * 状态变更回调
   */
  onStatusChange?: OnInteractionStatusChange;
  /**
   * 禁用交互
   */
  disabled?: boolean;
  /**
   * 响应画布式交互的父节点
   */
  getContainer?: GetContainer;
  /**
   * 前缀
   */
  prefixCls?: string;
  /**
   * 子组件
   */
  children?: ReactNode;
}
