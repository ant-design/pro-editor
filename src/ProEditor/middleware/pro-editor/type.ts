import { StateCreator } from 'zustand/vanilla';
import { ConfigPublicAction } from '../../store/slices/config';
import { TakeTwo, Write } from '../types/utils';

export interface ProEditorSetStateAction {
  type: unknown;

  /**
   * 如果记录，那么历史记录的名字是什么
   */
  name?: string;
  /**
   * 是否将操作记录到历史记录中
   */
  recordHistory?: boolean;
}

/**
 * 用于给注入方法添加额外的 proEditor 实例对象
 */
export interface ProEditorMiddlewareInjectMethod<Sa extends unknown[], Sr, T, A> {
  setState<A1 extends string | ProEditorSetStateAction>(...a: [...a: TakeTwo<Sa>, action?: A1]): Sr;

  proEditor: {
    undo?: () => void;
    redo?: () => void;
    // clearStorage: () => void;
    getOptions: () => Partial<ProEditorOptions<T, A>>;
  };
}
/**
 * 提供给用户的配置项
 */
export interface ProEditorOptions<S, EditorSaveState = S> {
  /** Name of the storage (must be unique) */
  name: string;
  /**
   * Filter the persisted value.
   *
   * @params state The state's value
   */
  partialize?: (state: S) => EditorSaveState;
}

// 为  mutator 注入 'pro-editor' 类型，以支持第三个配置参数

type WithProEditor<S, A> = S extends {
  getState: () => infer T;
  setState: (...a: infer Sa) => infer Sr;
}
  ? Write<S, ProEditorMiddlewareInjectMethod<Sa, Sr, T, A>>
  : never;

declare module 'zustand/vanilla' {
  interface StoreMutators<S, A> {
    ['pro-editor']: WithProEditor<S, A>;
  }
}

// 内部方法，用于保证中间件内部的类型定义安全

export type ProEditorImpl = <T>(
  storeInitializer: StateCreator<T, [['pro-editor', unknown]], []>,
  options: ProEditorOptions<T, T>,
) => StateCreator<T, [['pro-editor', unknown]], []>;

export interface InjectInternalProEditor {
  proEditor: {
    __INTERNAL_SET_CONFIG__NOT_USE_IT: ConfigPublicAction['setConfig'];
  };
}
