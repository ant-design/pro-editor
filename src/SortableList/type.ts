import type { UniqueIdentifier } from '@dnd-kit/core';
import type { ReactNode } from 'react';

export interface KeyManager {
  /**
   * 列表 ID 项
   */
  keys: UniqueIdentifier[];
  /**
   * 当前自增 ID
   */
  id: number;
}

export interface CreatorButtonProps {
  /**
   * 列表有值时是否展示添加按钮
   */
  showInList?: boolean;
  /**
   * 空数据时是否展示添加按钮
   */
  showInEmpty?: boolean;
  /**
   * 生成初始值逻辑
   */
  record?: (index: number) => Record<string, any>;
  /**
   * 新增一行按钮文案
   */
  creatorButtonText?: string;
}

export type RenderItemProps<T = any> = (item: T, index: number) => ReactNode;

export type RenderActionProps<T = any> =
  | ((item: T, index: number) => ReactNode[])
  | React.ReactNode[];

export type { UniqueIdentifier };
