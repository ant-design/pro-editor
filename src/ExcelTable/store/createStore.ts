import produce from 'immer';
import isEqual from 'lodash.isequal';
import { create } from 'zustand';

import { getNewRows, updateDataByColumn, updateDataByRow } from './utils';

import { initialState } from './initialState';

import type { CellChange } from 'handsontable/common';
import type { ExcelTableState, TableData } from './initialState';

/**
 * 组件交互点
 */
export interface ExcelTableAction {
  setStoreState: (nextState: Partial<ExcelTableState>) => void;
  /**
   * 内部更新 Data 方法
   * @param data
   */
  internalUpdateData: (data: TableData) => void;
  outsideUpdateData: (data: TableData) => void;
  setDefaultData: (data: TableData) => void;
  handleDataChanges: (changes: CellChange[]) => void;
  updateByColumn: (column: number) => void;
  updateByRow: (row: number) => void;
  resetData: () => void;
  reset: () => void;
}

export type ExcelTableStore = ExcelTableState & ExcelTableAction;

export const createStore = () =>
  create<ExcelTableStore>((set, get) => {
    return {
      ...initialState,
      setStoreState(nextState: Partial<ExcelTableState>) {
        set(nextState);
      },
      setDefaultData: (defaultData) => {
        set({ defaultData });
      },

      // 全量重置
      resetData: () => {
        const { defaultData, internalUpdateData } = get();
        internalUpdateData(defaultData);
      },

      reset: () => set({ ...initialState }),

      /**
       * 外部更新数据方法
       * @param data
       */
      outsideUpdateData: (data) => {
        set({ data });
      },
      /**
       * 内部更新 Data 方法
       * @param nextData
       */
      internalUpdateData: (nextData) => {
        const { data: prevData, onDataChange } = get();

        if (isEqual(prevData, nextData)) return;

        set({ data: nextData });

        // 受控更新
        onDataChange?.(nextData);
      },
      handleDataChanges: (changes) => {
        const { data, internalUpdateData } = get();

        const nextData = produce(data, (draft) => {
          changes.forEach((change) => {
            const [row, column, prev, next] = change;
            if (prev !== next) {
              // 如果 draft[row] 存在，说明是在行里面的
              if (draft[row]) {
                draft[row][column] = next;
              } else {
                // 否则说明行不够了，需要添加新行

                const diffRow = row + 1 - data.length;
                const newRows = getNewRows(diffRow, data[0].length);

                newRows.forEach((newRow) => {
                  draft.push(newRow);
                });

                draft[row][column] = next;
              }
            }
          });
        });

        internalUpdateData(nextData);
      },
      updateByColumn: (column) => {
        const newData = updateDataByColumn(column, get().data);

        get().internalUpdateData(newData);
      },
      updateByRow: (row) => {
        const newData = updateDataByRow(row, get().data);

        get().internalUpdateData(newData);
      },
    };
  });
