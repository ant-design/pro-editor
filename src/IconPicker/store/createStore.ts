import type { StoreApi } from 'zustand';
import { create } from 'zustand';
import { createContext, optionalDevtools } from 'zustand-utils';

import type { Store } from './store';
import vanillaStore from './store';

export const createStore = (showDevtools?: boolean) =>
  create<Store>()(optionalDevtools(showDevtools)(vanillaStore, { name: 'IconPicker' }));

export const { Provider, useStore, useStoreApi } = createContext<StoreApi<Store>>();
