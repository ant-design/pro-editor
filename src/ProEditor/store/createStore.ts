import type { StateCreator } from 'zustand';
import { create } from 'zustand';
import { optionalDevtools } from 'zustand-utils';

import {
  AwarenessPublicAction,
  AwarenessSlice,
  AwarenessSliceState,
  awarenessSlice,
} from './slices/awareness';
import { CanvasPublicAction, CanvasSlice, PublicCanvasState, canvasSlice } from './slices/canvas';
import { ConfigPublicAction, ConfigPublicState, ConfigSlice, configSlice } from './slices/config';
import { ConfigPanelSlice, configPanelSlice } from './slices/configPanel';
import {
  GeneralPublicAction,
  GeneralSlice,
  GeneralSliceState,
  generalSlice,
} from './slices/general';

/**
 * ProEditorState 接口描述编辑器状态
 * @template Config - 编辑器配置属性类型
 * @template PresenceAsset - 编辑器供外部感知的视图状态类型
 */
export type ProEditorState<Config = any, PresenceAsset = any> = GeneralSliceState &
  ConfigPublicState<Config> &
  PublicCanvasState &
  AwarenessSliceState<PresenceAsset>;

export type InternalProEditorStore = ProEditorState &
  ConfigSlice &
  CanvasSlice &
  ConfigPanelSlice &
  AwarenessSlice &
  GeneralSlice;

export type PublicProEditorStore = CanvasPublicAction &
  ConfigPublicAction &
  GeneralPublicAction &
  AwarenessPublicAction;

export type ProEditorStoreKey = keyof InternalProEditorStore;

const vanillaStore: StateCreator<InternalProEditorStore, [['zustand/devtools', never]]> = (
  ...params
) => ({
  ...generalSlice(...params),
  ...configSlice(...params),
  ...canvasSlice(...params),
  ...configPanelSlice(...params),
  ...awarenessSlice(...params),
});

export const createStore = (showDevTools: boolean = false) => {
  const devtools = optionalDevtools(showDevTools);

  return create<InternalProEditorStore>()(devtools(vanillaStore, { name: 'ProEditorStore' }));
};
