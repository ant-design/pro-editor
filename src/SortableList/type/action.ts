// 新增节点
export interface AddItemAction {
  /**
   * 动作类型
   */
  type: 'addItem';
  /**
   * 新增的节点
   */
  item: any;
  /**
   * 新增节点的位置
   * @default undefined
   */
  index?: number;
}

// 移动节点
export interface MoveItemAction {
  /**
   * 动作类型
   */
  type: 'moveItem';
  /**
   * 当前节点索引
   */
  activeIndex: number;
  /**
   * 目标节点索引
   */
  overIndex: number;
}

// 移除节点
export interface RemoveItemAction {
  /**
   * 动作类型
   */
  type: 'removeItem';
  /**
   * 要移除的节点的位置
   */
  index: number;
}

// 修改节点
export interface UpdateItemAction {
  /**
   * 动作类型
   */
  type: 'updateItem';
  /**
   * 要修改的节点的位置
   */
  index: number;
  /**
   * 修改后的节点内容
   */
  item: any;
}

/**
 * 内部 Item 数据更新方法
 */
export type SortableListDispatchPayload =
  | MoveItemAction
  | AddItemAction
  | RemoveItemAction
  | UpdateItemAction;
