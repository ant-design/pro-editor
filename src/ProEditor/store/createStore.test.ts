import { act, renderHook } from '@testing-library/react';
import { useEffect, useState } from 'react';

import type { CanvasInteractRule, InteractStatus } from '../../InteractContainer';
import { createStore } from './createStore';

vi.mock('zustand');

const useStore = createStore();

interface AssetConfig {
  data: any;
  columns: any;
}

describe('proEditorStore', () => {
  describe('状态值', () => {
    it('mode 默认为开发者模式', () => {
      const { result } = renderHook(() => useStore());
      expect(result.current.mode).toEqual('develop');
    });
  });
  describe('方法', () => {
    it('updatePosition', () => {
      const { result } = renderHook(() => useStore());

      expect(result.current.editorAwareness.panelPosition).toEqual({
        x: 0,
        y: 0,
      });

      act(() => {
        result.current.updatePanelPosition({ x: 12 });
      });
      expect(result.current.editorAwareness.panelPosition).toEqual({
        x: 12,
        y: 0,
      });

      act(() => {
        result.current.updatePanelPosition({ y: 1 });
      });
      expect(result.current.editorAwareness.panelPosition).toEqual({
        x: 12,
        y: 1,
      });
    });

    it('正确获取和设置 viewport', () => {
      const { result } = renderHook(() => useStore());

      act(() => {
        result.current.updateViewport({ zoom: 2 });
      });

      expect(result.current.editorAwareness.viewport).toEqual({
        zoom: 2,
        x: 0,
        y: 0,
      });
    });

    describe('修改配置', () => {
      it('默认', () => {
        const { result } = renderHook(() => useStore());

        expect(result.current.config).toEqual(null);

        act(() => {
          result.current.internalUpdateConfig({ hello: 'world' });
        });

        expect(result.current.config).toEqual({ hello: 'world' });
      });
      it('受控模式：没有 componentAssets 则不生成 props', () => {
        type Config = { text: string };
        const useTextStore = createStore();

        const { result } = renderHook(() => {
          const [value, onChange] = useState<Config>({ text: '' });
          const [p, onPropsChange] = useState();
          const store = useTextStore();

          useEffect(() => {
            useTextStore.setState({
              onConfigChange: ({ config, props }) => {
                onChange(config);
                onPropsChange(props);
              },
            });
            // eslint-disable-next-line react-hooks/exhaustive-deps
          }, []);

          useEffect(() => {
            if (!value) return;
            useTextStore.setState({ config: value });
            // eslint-disable-next-line react-hooks/exhaustive-deps
          }, [value]);

          return { store, value, props: p };
        });

        expect(result.current.store.config).toEqual({ text: '' });
        expect(result.current.props).toBeUndefined();

        act(() => {
          result.current.store.internalUpdateConfig({ text: '2' });
        });

        expect(result.current.store.config).toEqual({ text: '2' });
        expect(result.current.value).toEqual({ text: '2' });
        expect(result.current.props).toBeUndefined();
      });
      describe('受控模式，有 componentAssets', () => {
        const useTestStore = createStore();

        const useTestHook = () => {
          const [value, onChange] = useState<AssetConfig>();
          const [p, onPropsChange] = useState();
          const [empty, setIsEmpty] = useState<boolean>(true);
          const store = useTestStore();

          useEffect(() => {
            useTestStore.setState({
              onConfigChange: ({ config, isEmpty, props }) => {
                onChange(config as any);
                onPropsChange(props);
                setIsEmpty(isEmpty);
              },
            });
            // eslint-disable-next-line react-hooks/exhaustive-deps
          }, []);

          useEffect(() => {
            if (!value) return;
            useTestStore.setState({ config: value });
            // eslint-disable-next-line react-hooks/exhaustive-deps
          }, [value]);

          return {
            store,
            setConfig: onChange,
            value,
            props: p,
            isEmpty: empty,
          };
        };

        it('内部 config 更新时，内部 props 更新，同时外部值可接受到 props、config 和 isEmpty 值变化', () => {
          const { result } = renderHook(useTestHook);

          // 内部值
          expect(result.current.store.config).toBeNull();
          expect(result.current.store.props).toEqual({});

          // 外部值
          expect(result.current.props).toBeUndefined();
          expect(result.current.isEmpty).toBeTruthy();

          const config = {
            data: { dataType: 'oneApi' },
            columns: [{ dataIndex: 'a', title: 'hello' }],
          };

          act(() => {
            result.current.store.internalUpdateConfig(config);
          });

          // 内部值
          expect(result.current.store.config).toEqual(config);
          expect(result.current.store.props).toMatchSnapshot();

          // 外部值
          expect(result.current.value).toEqual(config);
          expect(result.current.isEmpty).toBeFalsy();
          expect(result.current.props).toMatchSnapshot();
        });

        it('外部设置 config 时，内部 props 会跟随 config 更新，不触发 onChange', () => {
          const { result } = renderHook(useTestHook);

          expect(result.current.store.config).toBeNull();
          expect(result.current.store.props).toEqual({});
          expect(result.current.isEmpty).toBeTruthy();

          const config = {
            data: { dataType: 'oneApi' },
            columns: [{ dataIndex: 'a', title: 'hello' }],
          };

          act(() => {
            result.current.setConfig(config);
          });

          // 内部值
          expect(result.current.store.config).toEqual(config);
          expect(result.current.store.props).toMatchSnapshot();

          // 外部值
          expect(result.current.value).toEqual(config);
          expect(result.current.isEmpty).toBeTruthy();
          expect(result.current.props).toBeUndefined();
        });
      });
    });

    describe('画布交互', () => {
      it('默认', () => {
        const rule: CanvasInteractRule = {
          id: 'column',
          selectors: [],
          actions: ['click'],
        };
        const { result } = renderHook(() => useStore());

        expect(result.current.interaction).toEqual({});

        act(() => {
          result.current.internalUpdateCanvasInteract({
            action: 'click',
            currentNode: { rule, index: 2 },
          });
        });

        expect(result.current.interaction).toEqual({
          action: 'click',
          currentNode: { index: 2, rule },
        });

        act(() => {
          result.current.internalUpdateCanvasInteract({
            status: 'unSelected',
          });
        });

        expect(result.current.interaction).toEqual({
          status: 'unSelected',
        });
      });
      it('受控模式', () => {
        const rule: CanvasInteractRule = {
          id: 'column',
          selectors: [],
          actions: ['click'],
        };
        const { result } = renderHook(() => {
          const [value, onChange] = useState<InteractStatus>({
            action: 'hover',
          });
          const store = useStore();

          useEffect(() => {
            useStore.setState({ onInteractionChange: onChange });
            // eslint-disable-next-line react-hooks/exhaustive-deps
          }, []);

          useEffect(() => {
            if (!value) return;
            useStore.setState({ interaction: value });
            // eslint-disable-next-line react-hooks/exhaustive-deps
          }, [value]);

          return { store, value };
        });

        expect(result.current.store.interaction).toEqual({ action: 'hover' });

        act(() => {
          result.current.store.internalUpdateCanvasInteract({
            action: 'click',
            currentNode: { rule, index: 2 },
          });
        });

        expect(result.current.store.interaction).toEqual({
          action: 'click',
          currentNode: { index: 2, rule },
        });

        expect(result.current.value).toEqual({
          action: 'click',
          currentNode: { index: 2, rule },
        });
      });
      it('toggleCanvasInteraction', () => {
        const { result } = renderHook(() => useStore());

        expect(result.current.enableCanvasInteraction).toBeTruthy();

        act(() => {
          result.current.toggleCanvasInteraction();
        });

        expect(result.current.enableCanvasInteraction).toBeFalsy();
        act(() => {
          result.current.toggleCanvasInteraction();
        });

        expect(result.current.enableCanvasInteraction).toBeTruthy();
      });

      it('取消选中', () => {
        const rule: CanvasInteractRule = {
          id: 'column',
          selectors: [],
          actions: ['click'],
        };
        const { result } = renderHook(() => useStore());

        expect(result.current.interaction).toEqual({});

        act(() => {
          result.current.internalUpdateCanvasInteract({
            action: 'click',
            currentNode: { rule, index: 3 },
          });
        });

        expect(result.current.interaction).toEqual({
          action: 'click',
          currentNode: { index: 3, rule },
        });

        act(() => {
          result.current.deselectCanvas();
        });

        expect(result.current.interaction).toEqual({
          status: 'unSelected',
        });
      });
    });
  });
});
