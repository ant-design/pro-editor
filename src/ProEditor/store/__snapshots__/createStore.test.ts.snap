// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`proEditorStore > 方法 > 修改配置 > 受控模式，有 componentAssets > 内部 config 更新时，内部 props 更新，同时外部值可接受到 props、config 值变化 1`] = `{}`;

exports[`proEditorStore > 方法 > 修改配置 > 受控模式，有 componentAssets > 内部 config 更新时，内部 props 更新，同时外部值可接受到 props、config 值变化 2`] = `undefined`;

exports[`proEditorStore > 方法 > 修改配置 > 受控模式，有 componentAssets > 外部设置 config 时，内部 props 会跟随 config 更新，不触发 onChange 1`] = `{}`;
